### DISCUSSION TITLE
Smart Intercom using ESP32

### DISCUSSION THREAD

**Discussion by Philomath88** [pschatzmann/arduino-audio-tools#1620] [Q&A] [answered]
Hey guys,

am trying to make my STR 200 TV intercom smart. 

I am currently working on picking up the microphone audio (which is the easiest since I don't need another person at the door). 

The output is 4.3V so I build a voltage divider circuit using resistors (the probe divides by 10): 

This is in theory my script to output the audio signal to a web browser (just for testing): https://youtu.be/rurvSKYao3U

`

#include "AudioTools.h"
#include <WiFi.h>
#include "driver/i2s.h"
#include "driver/adc.h"

// WIFI
const char* ssid = "gigacube-46F37C";
const char* password = "7fLf2t5t5j233839";

// Static IP configuration
IPAddress staticIP(192, 168, 1, 200);  // The static IP you want to assign to your ESP32
IPAddress gateway(192, 168, 1, 1);     // Your router's IP address
IPAddress subnet(255, 255, 255, 0);    // Subnet mask
IPAddress dns(8, 8, 8, 8);             // DNS (you can use Google's public DNS)

AudioWAVServer server(ssid, password);

// ADC and I2S Configuration
#define SAMPLE_RATE 16000  // 16kHz
#define I2S_NUM I2S_NUM_0
#define DMA_BUF_COUNT 8
#define DMA_BUF_LEN 512

class ADCStream : public AudioStream {
public:
    ADCStream() : AudioStream() {
        // I2S configuration
        i2s_config_t i2s_config = {
            .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_RX | I2S_MODE_ADC_BUILT_IN),
            .sample_rate = SAMPLE_RATE,
            .bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT,
            .channel_format = I2S_CHANNEL_FMT_ONLY_RIGHT,
            .communication_format = I2S_COMM_FORMAT_I2S_MSB,
            .intr_alloc_flags = ESP_INTR_FLAG_LEVEL1,
            .dma_buf_count = DMA_BUF_COUNT,
            .dma_buf_len = DMA_BUF_LEN,
            .use_apll = false,
            .tx_desc_auto_clear = false,
            .fixed_mclk = 0
        };

        // Install and start I2S driver
        if (i2s_driver_install(I2S_NUM, &i2s_config, 0, NULL) != ESP_OK) {
            Serial.println("Failed to install I2S driver");
            return;
        }

        // Set I2S ADC mode
        if (i2s_set_adc_mode(ADC_UNIT_1, ADC1_CHANNEL_4) != ESP_OK) {
            Serial.println("Failed to set I2S ADC mode");
            return;
        }

        // Enable ADC
        if (i2s_adc_enable(I2S_NUM) != ESP_OK) {
            Serial.println("Failed to enable I2S ADC");
            return;
        }
    }

    size_t readBytes(uint8_t *data, size_t len) override {
        size_t bytes_read;
        if (i2s_read(I2S_NUM, data, len, &bytes_read, portMAX_DELAY) == ESP_OK) {
            return bytes_read;
        }
        return 0;
    }
};

ADCStream adcStream;

void setup() {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Info);

    // Configure static IP
    if (!WiFi.config(staticIP, gateway, subnet, dns)) {
        Serial.println("STA Failed to configure");
    }

    // Connect to Wi-Fi
    WiFi.begin(ssid, password);
    while (WiFi.status() != WL_CONNECTED) {
        delay(1000);
        Serial.println("Connecting to WiFi...");
    }

    // Print ESP32 IP address
    Serial.println("Connected to WiFi");
    Serial.print("IP address: ");
    Serial.println(WiFi.localIP());

    // start server
    server.begin(adcStream, SAMPLE_RATE, 1);  // 1 channel

    Serial.println("ADC stream server started");
}

void loop() {
    server.copy();
}
`

Unfortunately, I don't hear anything. I tried outputting the ADC values and if I speak, the go up to a max of 300. 

Does anyone have an idea what I am doing wrong?

**Comment by pschatzmann** [comment, marked as answer]
Why adding all this complexity ? Just use an AnalogAudioStream in input mode!
Here is an [example that reads from an analog mocrophone](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-adc-serial)...

ps. your code will not work anymore if you upgrade to an actual ESP32 core version because the I2S API has changed!

**Comment by Philomath88** [comment]
@pschatzmann Thank you for the fast reply! What code should I use? I am using an ESP32-WROOM-32 board. 

The ultimate goal is to send the audio data via a web socket to my server

**Comment by Philomath88** [comment]
I have updated my code to this: 

`#include "Arduino.h"
#include "AudioTools.h"
#include <WiFi.h>

// WIFI credentials
const char* ssid = "gigacube-46F37C";
const char* password = "7fLf2t5t5j233839";

// Static IP configuration
IPAddress staticIP(192, 168, 1, 200);
IPAddress gateway(192, 168, 1, 1);
IPAddress subnet(255, 255, 255, 0);
IPAddress dns(8, 8, 8, 8);

// TCP settings
WiFiServer server(8888);
WiFiClient client;

AnalogAudioStream in;
AudioInfo info(8000, 1, 16);

unsigned long sampleCount = 0;

void setup() {
    Serial.begin(115200);
    
    auto cfgRx = in.defaultConfig(RX_MODE);
    cfgRx.adc_pin = 32;
    cfgRx.copyFrom(info);
    in.begin(cfgRx);

    // Configure static IP
    if (!WiFi.config(staticIP, gateway, subnet, dns)) {
        Serial.println("STA Failed to configure");
    }

    // Connect to WiFi
    WiFi.begin(ssid, password);
    while (WiFi.status() != WL_CONNECTED) {
        delay(1000);
        Serial.println("Connecting to WiFi...");
    }
    Serial.println("Connected to WiFi");
    Serial.print("IP address: ");
    Serial.println(WiFi.localIP());

    server.begin();
    Serial.println("TCP server started");
}

void loop() {
    if (!client.connected()) {
        client = server.available();
        if (client) {
            Serial.println("New client connected");
        }
    }

    if (client.connected()) {
        int16_t sample;
        if (in.readBytes((uint8_t*)&sample, sizeof(sample)) == sizeof(sample)) {
            char buffer[50];
            snprintf(buffer, sizeof(buffer), "%lu,%d\n", sampleCount, sample);
            client.print(buffer);
            sampleCount++;
        }
    }

    // Small delay to prevent overwhelming the system
    delay(1);
}`

And am using this python script to save the values: 

import socket
import csv
from datetime import datetime

TCP_IP = "192.168.1.200"  # ESP32's IP address
TCP_PORT = 8888

# Generate a unique filename based on current timestamp
filename = f"audio_data_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.connect((TCP_IP, TCP_PORT))
    print(f"Connected to {TCP_IP}:{TCP_PORT}")

    with open(filename, 'w', newline='') as csvfile:
        csvwriter = csv.writer(csvfile)
        csvwriter.writerow(['Sample', 'Value'])  # Write header
        
        try:
            buffer = ""
            while True:
                data = s.recv(1024).decode('ascii')
                buffer += data
                lines = buffer.split('\n')
                for line in lines[:-1]:
                    sample, value = line.strip().split(',')
                    print(f"Received: Sample {sample}, Value {value}")
                    csvwriter.writerow([sample, value])
                buffer = lines[-1]
                csvfile.flush()  # Ensure data is written to file immediately
        except KeyboardInterrupt:
            print("Recording stopped by user")

print(f"Data saved to {filename}")


[audio_data_20240716_172132.csv](https://github.com/user-attachments/files/16252670/audio_data_20240716_172132.csv)


Is this what I should expect? Given that the above code is not working anymore for the board, what approach would you recommend so I can broadcast the digitised audio to web to listen to it (just to verify it is working)?

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

Just test the value range of the ADC signal first with the sketch that I have sent you!
In the optimum case the voltage should oszillate between 0 and 3.3V which would give values from - 32768 to +32767

**Comment by Philomath88** [comment]
@pschatzmann I have set it to D32 pin. This is the serial output: 

https://pastebin.com/FTTLyiFW

What i have seen in the Oscilloscope is that it hovers around 2.1V and then fluctuates by a maximum of 0.1V

**Comment by pschatzmann** [comment]
Just record a sine wave and display it with the Arduino Serial Plotter.
This will give you a better feeling if you get a proper signal that you can postprocess with a ConverterScaler (shift and multiply)

ps: send out the raw data and convert it on the desktop. 
Sending ascii is adding too much overhead....
